src/coreclr/debug/createdump/crashinfo.cpp

```diff
@@ -683,72 +683,95 @@ CrashInfo::InsertMemoryRegion(uint64_t address, size_t size)
     }
-    // Either part of the region was invalid, part of it hasn't been added or the backed
-    // by memory state is different.
-    uint64_t start = region.StartAddress();
 
-    // The region overlaps/conflicts with one already in the set so add one page at a
-    // time to avoid the overlapping pages.
-    uint64_t numberPages = region.SizeInPages();
+    // Go page by page and split the region into valid sub-regions
+    uint64_t pageStart = region.StartAddress();
+    uint64_t numberPages = region.Size() / PAGE_SIZE;
+    uint64_t subRegionStart, subRegionEnd;
     int pagesAdded = 0;
-
-    for (size_t p = 0; p < numberPages; p++, start += PAGE_SIZE)
+    subRegionStart = subRegionEnd = pageStart;
+    for (size_t p = 0; p < numberPages; p++, pageStart += PAGE_SIZE)
     {
-        MemoryRegion memoryRegionPage(region.Flags(), start, start + PAGE_SIZE);
+        MemoryRegion page(region.Flags(), pageStart, pageStart + PAGE_SIZE);
 
-        const auto& found = m_memoryRegions.find(memoryRegionPage);
-        if (found == m_memoryRegions.end())
+        // avoid searching for conflicts if we know we don't have one
+        const bool pageHasConflicts = hasConflict && m_memoryRegions.find(page) != m_memoryRegions.end();
+        // avoid validating the page if it conflicts: we won't add it in any case
+        const bool pageIsValid = !pageHasConflicts && PageMappedToPhysicalMemory(pageStart) && PageCanBeRead(pageStart);
+
+        if (pageIsValid)
         {
-            // All the single pages added here will be combined in CombineMemoryRegions()
-            if (ValidRegion(memoryRegionPage))
+            subRegionEnd = page.EndAddress();
+            pagesAdded++;
+        }
+        else
+        {
+            // the next page is not valid thus sub-region is complete
+            if (subRegionStart != subRegionEnd)
             {
-                m_memoryRegions.insert(memoryRegionPage);
-                pagesAdded++;
+                m_memoryRegions.insert(MemoryRegion(region.Flags(), subRegionStart, subRegionEnd));
             }
+            subRegionStart = subRegionEnd = page.EndAddress();
         }
     }
+    // add the last sub-region if it's not empty
+    if (subRegionStart != subRegionEnd)
+    {
+        m_memoryRegions.insert(MemoryRegion(region.Flags(), subRegionStart, subRegionEnd));
+    }
 
     return pagesAdded;
 }
 
-//
-// Validates a memory region
-//
 bool
-CrashInfo::ValidRegion(const MemoryRegion& region)
+CrashInfo::PageMappedToPhysicalMemory(uint64_t start)
 {
-    uint64_t start = region.StartAddress();
-    uint64_t numberPages = region.SizeInPages();
-    for (size_t p = 0; p < numberPages; p++, start += PAGE_SIZE)
-    {
-        BYTE buffer[1];
-        size_t read;
+    #ifdef __APPLE__
+        // this check has not been implemented yet for macos
+        return true;
+    #else
+        // https://www.kernel.org/doc/Documentation/vm/pagemap.txt
+        if (m_fdPagemap == -1)
```

Comment:
I checked and these bits only require `PTRACE_MODE_READ_FSCREDS`, which createdump already uses, so it shouldn't really cause any issues.